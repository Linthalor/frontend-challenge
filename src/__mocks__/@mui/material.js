const genMockComponent = (name) => (props) => {
  const MockName = `mocked-mui-${name}`;
  const {sx, ...remainingProps} = props;
  return <MockName {...remainingProps} sx={sx ? JSON.stringify(sx) : undefined} />;
}

module.exports = {
  ...jest.requireActual('@mui/material'),
  Box: genMockComponent('box'),
  Button: genMockComponent('button'),
  IconButton: genMockComponent('icon-button'),
  InputAdornment: genMockComponent('input-adornment'),
  TextField: genMockComponent('text-field'),
  Typography: genMockComponent('typography'),
  Card: genMockComponent('card'),
  CardContent: genMockComponent('card-content'),
  CardActions: genMockComponent('card-actions'),
};