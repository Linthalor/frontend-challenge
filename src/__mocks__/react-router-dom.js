module.exports = {
  ...jest.requireActual('react-router-dom'),
  Navigate: (props) => {
    const MockName = `mocked-react-router-dom-navigate`;
    return <MockName {...props} />;
  }
};