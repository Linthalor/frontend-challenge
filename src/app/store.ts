import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { signUpReducer } from '../features/sign-up/sign-up.slice';
import { colorReducer } from '../features/color/color.slice';

export const store = configureStore({
  reducer: {
    signUp: signUpReducer,
    color: colorReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
