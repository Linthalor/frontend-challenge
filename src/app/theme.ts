import { createTheme } from '@mui/material';

// Themed using Upgrade's palette
export const theme = createTheme({
  palette: {
    primary: {
      light: '#60c7a1',
      main: '#00a669',
      dark: '#008051',
    },
    secondary: {
      main: '#3f3d56',
      light: '#6bc1fe',
    },
    success: {
      main: '#03b57a',
      contrastText: '#ffffff',
    },
    error: {
      main: '#c83c4c',
      contrastText: '#ffffff',
    },
    text: {
      primary: '#3f465e',
      secondary: '#008051'
    }
  },
});