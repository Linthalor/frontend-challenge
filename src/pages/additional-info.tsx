import {
  Box,
  Button,
  Checkbox,
  CircularProgress,
  FormControl,
  FormControlLabel,
  FormHelperText,
  InputAdornment,
  InputLabel,
  MenuItem,
  Select,
  Skeleton,
  Typography,
} from '@mui/material';
import { useEffect, useRef } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { useAppDispatch } from '../app/hooks';
import { SignUpCard } from '../components/signup-card';
import { selectUserPrefsValues, setUserPrefsValues } from '../features/sign-up/sign-up.slice';
import { SignUpUserPrefsValues, SignUpUserPrefsValuesSchema } from '../model/sign-up';
import { useSelector } from 'react-redux';
import { isError, isLoading, isValue } from '../model/util/async-prop';
import { TermsAndConditions, TermsAndConditionsRef } from '../components/terms-and-conditions';
import { useTranslation } from 'react-i18next';
import { Controller, useForm, } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { getColorsAsync, selectColors } from '../features/color/color.slice';

export const AdditionalInfo = () => {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { t } = useTranslation();
  const colors = useSelector(selectColors);
  const termsRef = useRef<TermsAndConditionsRef | null>(null);
  const signUpPrefsData = useSelector(selectUserPrefsValues);

  useEffect(() => {
    /*
      This could potentially be moved to a higher level so it fetches at app load and
      not every time this section of the form is visited, or use a caching strategy
      to prevent reloading it if it's not been updated.
      
      For the sake of demoing the async nature of colors, it will be left as is.
     */
    dispatch(getColorsAsync());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const { handleSubmit, register, formState, control, watch } = useForm({
    resolver: zodResolver(SignUpUserPrefsValuesSchema),
    mode: 'onChange',
    defaultValues: {
      color: signUpPrefsData?.color || '',
      terms: signUpPrefsData?.terms || false,
    },
  });
  const { errors } = formState;

  const submitForm = (values: SignUpUserPrefsValues) => {
    dispatch(setUserPrefsValues(values));
    navigate('/confirmation');
  }

  return (
    <form
      style={{
        height: '100%',
        width: '100%',
      }}
      onSubmit={handleSubmit(submitForm)}
    >
      <SignUpCard
        content={<>
          <Typography
            sx={{
              textAlign: 'center',
              fontWeight: 'bold'
            }}
            variant="h5"
            color="text.primary"
            gutterBottom
          >
            {t('page:additionalInfo:label')}
          </Typography>
          <Controller
            name="color"
            control={control}
            render={({ field }) => (
              <FormControl error={!!errors.color?.message}>
                <InputLabel id="favorite-color-select-field-label" required>{t('form:color:label')}</InputLabel>
                <Select
                  id="favorite-color-select-field"
                  label={t('form:color:label')}
                  required
                  endAdornment={isLoading(colors) && (
                    <InputAdornment position="end" sx={{ mr: '-7px' }}>
                      <CircularProgress size={24} />
                    </InputAdornment>
                  )}
                  {...field}
                >
                  {isLoading(colors) && <MenuItem value={watch('color')}>{watch('color')}</MenuItem>}
                  {isLoading(colors) && [1,2,3,4,5].map(id => 
                    <MenuItem key={id} disabled>
                      <Skeleton variant='text' sx={{width: '100%'}}/>
                    </MenuItem>
                  )}
                  {isError(colors) && <MenuItem disabled>{t('form:color:errors:failed_async')}</MenuItem>}
                  {isValue(colors) && colors.value.map(color =>
                    <MenuItem key={color} value={color}>{color}</MenuItem>
                  )}
                </Select>
                <FormHelperText>{errors.color?.message}</FormHelperText>
              </FormControl>
            )}
          />
          <FormControlLabel
            control={
              <Checkbox {...register('terms')} checked={watch('terms')} />
            }
            label={
              <Box
                sx={{
                  display: 'flex',
                  flexDirection: 'row',
                  alignItems: 'center'
                }}
              >
                <Typography>{t('common:terms:agree')}</Typography>
                <Button
                  onClick={() => termsRef.current?.setOpenTerms(true)}
                >
                  {t('common:terms:termsAndConditions')}
                </Button>
              </Box>
            }
          />
        </>}
        actions={<>
          <Button variant="outlined" color="secondary" component={Link} to="/sign-up">
            {t('common:back')}
          </Button>
          <Button variant="contained" type="submit" disabled={!formState.isValid}>
            {t('common:next')}
          </Button>
        </>}
      />
      <TermsAndConditions ref={termsRef}/>
    </form>
  );
};