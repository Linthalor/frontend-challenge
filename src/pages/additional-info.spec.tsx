import { cleanup, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createTestStore, store as realStore } from '../app/store';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '@mui/material';
import { theme } from '../app/theme';
import { AdditionalInfo } from './additional-info';

jest.mock('@mui/material');

describe('AdditionalInfo', () => {
  let store: typeof realStore;
  beforeEach(() => {
    store = createTestStore();
  });
  afterEach(() => {
    cleanup();
  });

  it('renders as expected', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <BrowserRouter>
            <AdditionalInfo />
          </BrowserRouter>
        </Provider>
      </ThemeProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});