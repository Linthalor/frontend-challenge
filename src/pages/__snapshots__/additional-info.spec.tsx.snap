// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalInfo renders as expected 1`] = `
<form
  style="height: 100%; width: 100%;"
>
  <mocked-mui-box
    sx="{\\"width\\":\\"100%\\",\\"height\\":\\"100%\\",\\"display\\":\\"flex\\",\\"justifyContent\\":\\"center\\",\\"alignItems\\":\\"center\\"}"
  >
    <mocked-mui-card
      sx="{\\"width\\":400,\\"height\\":\\"unset\\",\\"display\\":\\"flex\\",\\"flexDirection\\":\\"column\\",\\"justifyContent\\":\\"center\\"}"
    >
      <mocked-mui-card-content
        sx="{\\"display\\":\\"flex\\",\\"flexDirection\\":\\"column\\",\\"alignItems\\":\\"stretch\\",\\"justifyContent\\":\\"start\\",\\"gap\\":1,\\"flex\\":1}"
      >
        <mocked-mui-typography
          color="text.primary"
          gutterbottom="true"
          sx="{\\"textAlign\\":\\"center\\",\\"fontWeight\\":\\"bold\\"}"
          variant="h5"
        >
          Additional Info
        </mocked-mui-typography>
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-zq33gc-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            id="favorite-color-select-field-label"
          >
            Favorite Color
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-1o0vv86-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd  css-aucmf9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            required=""
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="favorite-color-select-field"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-4ygxeq-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="favorite-color-select-field"
              role="button"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              name="color"
              required=""
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <mocked-mui-input-adornment
              position="end"
              sx="{\\"mr\\":\\"-7px\\"}"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-172ucwx-MuiCircularProgress-root"
                role="progressbar"
                style="width: 24px; height: 24px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </mocked-mui-input-adornment>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Favorite Color
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-12kfrt5-MuiFormHelperText-root"
          />
        </div>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-1uxrgnw-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              name="terms"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            <mocked-mui-box
              sx="{\\"display\\":\\"flex\\",\\"flexDirection\\":\\"row\\",\\"alignItems\\":\\"center\\"}"
            >
              <mocked-mui-typography>
                I Agree to the
              </mocked-mui-typography>
              <mocked-mui-button>
                Terms & Conditions
              </mocked-mui-button>
            </mocked-mui-box>
          </span>
        </label>
      </mocked-mui-card-content>
      <mocked-mui-card-actions
        sx="{\\"display\\":\\"flex\\",\\"flexDirection\\":\\"row\\",\\"justifyContent\\":\\"end\\",\\"gap\\":1}"
      >
        <mocked-mui-button
          color="secondary"
          component="[object Object]"
          to="/sign-up"
          variant="outlined"
        >
          Back
        </mocked-mui-button>
        <mocked-mui-button
          disabled="true"
          type="submit"
          variant="contained"
        >
          Next
        </mocked-mui-button>
      </mocked-mui-card-actions>
    </mocked-mui-card>
  </mocked-mui-box>
</form>
`;
